File: Nodes/TreapNode/treap_node.js
/**
 * Generates a random integer between two inclusive bounds.
 *
 * @param {number} min - The minimum value (inclusive).
 * @param {number} max - The maximum value (inclusive).
 * @returns {number} A random integer between min and max (inclusive).
 */
function getRandomInt(min, max) {
    min = Math.ceil(min);  // Round up min to the nearest integer
    max = Math.floor(max); // Round down max to the nearest integer
    return Math.floor(Math.random() * (max - min + 1)) + min; // Random integer
}

/**
 * Represents a node in the Treap Tree.
 *
 * A TreapNode is a combination of a binary search tree (key ordering) and a max-heap (priority ordering).
 * Each node contains a key, a randomly assigned priority (or a provided one), and pointers to left and right children.
 *
 * @class
 */
class TreapNode {
    /**
     * Creates a TreapNode instance.
     *
     * @param {number} key - The key of the node.
     * @param {number|null} [priority=null] - The priority of the node.
     *                                        If not provided, a random priority between 1 and 100 is assigned.
     */
    constructor(key, priority = null) {
        this.key = key;
        this.priority = priority ? priority : getRandomInt(1, 100); // Assign random priority if not provided
        this.left = null;  // Pointer to the left child
        this.right = null; // Pointer to the right child
    }
}

File: Nodes/TreapNode/treap_node.py
import random


class TreapNode:
    """
    A node in the Treap Tree.

    A TreapNode combines properties of both a binary search tree and a heap:
    - The key maintains the binary search tree property (keys in the left subtree are smaller, and keys in the right subtree are larger).
    - The priority maintains the heap property (a parent's priority is higher than its children).

    Attributes:
        key (int): The key value of the node.
        priority (int): The priority of the node, used to maintain the heap property.
                        If not provided, a random priority is assigned.
        left (TreapNode or None): Pointer to the left child.
        right (TreapNode or None): Pointer to the right child.
    """

    def __init__(self, key, priority=None):
        """
        Initialize a TreapNode with a given key and optional priority.

        If the priority is not provided, a random priority is generated between 1 and 100.

        Args:
            key (int): The key value for the node.
            priority (int, optional): The priority value for the node. Defaults to a random value.

        Example:
            >>> node = TreapNode(10)
            >>> print(node.key, node.priority)  # Key is 10, priority is a random value
        """
        self.key = key
        self.priority = priority if priority is not None else random.randint(1, 100)
        self.left = None
        self.right = None
